Adapt alarm notifications management to new API

dpino opened the pull request:
Command `alarms listen` and management of alarm notification in `ptree.lua` need to be update to the new fibers API.

It's not fully working yet. As soon as I receive a notification, I got an EOF error.

Here's what I have done:
* In `alarms listen` I have only one fiber that pulls messages from the leader. In `config listen` there are two fibers, one for request and another one for replies. The request fiber puts replies into a queue, which is later pulled by the reply fiber. I suspect that I need a similar mechanism, but how can I know when a notification has been written in the leader?
* In the management of notifications in ptree.lua, before notifications were queued into a peer and later pulled and sent one by one. There was tracking information of the last position written of the first item. I'm not sure if this mechanism is still needed.

dpino committed:
Adapt alarm notifications management to new API

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -602,55 +594,31 @@ end
+            local msg = peer.queue[1]
+            peer:write_chars(msg)
+            peer:flush_output()
+            table.remove(peer.queue, 1)
```
I think you don't want to spawn a fiber here.  With fibers, generally what you want to do is to spawn one fiber per peer:

```lua
function Manager:send_notifications_to_peer(peer)
   self:call_with_cleanup(peer, function()
      while true do
         local msg = notification_queue:get()
         peer:write_chars(msg)
         peer:flush_output()
      end
   end)
end
```

i.e. the fiber lives as long as the peer is connected.  So what about the queue?  The peculiar thing is that it needs to cause the fiber to block when the queue is empty, and only wake up the fiber when something is added to the queue.  For that I would usually recommend a concurrent ML queue; see `lib.fibers.queue`.  A queue in CML is really just a pair of channels, with a dedicated fiber that does the queue functionality.

In your code, you were spawning a fiber per notification.  In the case where writing the notification takes time, you could have two fibers writing on the same socket!  Additionally after writing, the first finishing fiber closed the socket -- that's probably where your EOF comes from.

Regarding queues, I do have a couple of doubts.  One is that really you can only push onto a queue from a fiber.  So you'd need to have a fiber polling for alarms.notifications(), and then putting items on each peer's queue.  That's fine:

```lua
-- spawn in a fiber
function Manager:notification_poller()
   while true do
      if #alarms.notifications == 0 then
         fiber.sleep(1/50) -- poll at 50 Hz
      else
         -- here we enqueue the notifications to peer fibers
      end
   end
end
```

Then you remove `push_notifications_to_peers`, as running the scheduler will run the polling fiber.  That's nice.

However my doubt is about the queue.  The way fibers works currently is that when a fiber causes another fiber to be runnable, that fiber is enqueued to run on the next turn.  Similarly, if a fiber blocks, the soonest it could run again would be the next turn.  So when a queue is empty and you push a message to it, you wake up the queue fiber and you continue.  If you try to push another message, you will block, because the queue's fiber isn't ready to receive another value yet -- it needs to run first.  Even though the queue is asynchronous, it takes turns to communicate values.  So if you have a lot of notifications coming through, their dispatch rate could be limited by the breath rate, if each breath runs one turn.

There are various ways we could solve this.  We could resume fibers into the runqueue of the current turn, but that could lead to the process devoting proportionally too much time to fibers, when maybe it has some other things it needs to do as well.  Not sure.  It could be of course that this isn't a problem at all!

dpino committed:
Add identificator name in alarms

There are 3 types of alarms according to the alarms schema.  In the case
of alarm-inventory-changed notifications, there is no fields,  so it's
hard to figure out what type of alarm is that.  This patchs adds to the
notifications an extra field called 'event' with the name of the
notification.

dpino commented:
@wingo: It's working now, thanks for the tips. However, it's not fully working yet. When there are multiple peers connected, only one gets notified. I will look into that.

Another issue I noticed is that messages get enqueued and when the first peer gets connected all the messages are flushed. For instance, when the lwAFTR is started and the alarm-inventory is created, several 'alarm-inventory-changed' notifications are emitted (this is the expected behavior according to the spec). To me if there are not listeners connected at that time, those notifications should get lost (this apply to any other type of notification emitted when there are no listeners connected).

On the other hand, is not that bad seeing these notifications as there might be an elapsing time since a lwAFTR instance is started until a peer connects to it. For instance, having these notifications enqueued helps to check whether there was a notification while starting the lwAFTR.

Example on start:

```bash
$ sudo alarms listen lwaftr 
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"alarm_text":"Make sure you can resolve external-interface.next-hop.ip address manually.  If it cannot be resolved, consider setting the MAC address of the next-hop directly.  To do it so, set external-interface.next-hop.mac to the value of the MAC address.","perceived_severity":"critical","time":"2018-05-10T07:34:58Z","event":"alarm-notification"}
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
{"event":"alarm-inventory-changed"}         
```

wingo reviewed (approved):
very nice!
